client/client.go

```diff
@@ -0,0 +1,150 @@
+package client
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"time"
+
+	"github.com/google/netstack/tcpip"
+	"github.com/google/netstack/tcpip/header"
+	"github.com/u-root/dhcp6"
+)
+
+const (
+	srcPort = 546
+	dstPort = 547
+)
+
+var (
+	// All DHCP servers and relay agents on the local network segment (RFC 3315)
+	// IPv6 Multicast (RFC 2464)
+	// insert the low 32 Bits of the multicast IPv6 Address into the Ethernet Address (RFC 7042 2.3.1.)
+	multicastMAC = net.HardwareAddr([]byte{0x33, 0x33, 0x00, 0x01, 0x00, 0x02})
+)
+
+type Client struct {
+	// The HardwareAddr to send in the request.
+	srcMAC net.HardwareAddr
+
+	// Packet socket to send on.
+	connection *packetSock
+
+	// Timeout
+	timeout time.Duration
+
+	// Max number of attempts to send DHCPv6 solicit to server.
+	// A valid DHCPv6 reply is supposed to be received by client before timeout.
+	// -1 means infinity.
+	retry int
+}
+
+func New(haddr net.HardwareAddr, packetSock *packetSock, t time.Duration, r int) *Client {
+	return &Client{
+		srcMAC:     haddr,
+		connection: packetSock,
+		timeout:    t,
+		retry:      r,
+	}
+}
+
+func (c *Client) Solicit() ([]*dhcp6.IAAddr, *dhcp6.Packet, error) {
+	solicitPacket, err := newSolicitPacket(c.srcMAC)
+	if err != nil {
+		return nil, nil, fmt.Errorf("new solicit packet: %v", err)
+	}
+
+	var packet *dhcp6.Packet
+	for i := 0; i < c.retry || c.retry < 0; i++ { // each retry takes the amount of timeout at worst.
+		if err := c.SendPacket(solicitPacket); err != nil {
+			return nil, nil, fmt.Errorf("send solicit packet(%v) = err %v", solicitPacket, err)
+		}
+
+		packet, err = c.ReadReply()
+		if err != nil {
+			log.Printf("%v\nResending DHCPv6 Solicit Message...", err)
```

Comment:
If we are going to log at all, I'd like the Client to expose a logger field so it doesn't just always go to stdout/stderr. A library shouldn't log with a global IMO.